src/circuit/gadgets/constraints.rs

```diff
@@ -73,32 +94,48 @@ pub(crate) fn add<F: PrimeField, CS: ConstraintSystem<F>>(
     Ok(res)
 }
 
+/// Creates a linear combination representing the popcount (sum of one bits) of `v`.
+pub(crate) fn popcount_lc<F: PrimeField, CS: ConstraintSystem<F>>(
+    v: &[Boolean],
+) -> Result<LinearCombination<F>, SynthesisError> {
+    v.iter()
+        .try_fold(LinearCombination::<F>::zero(), |acc, bit| {
+            add_to_lc::<F, CS>(bit, acc, F::ONE)
+        })
+}
+
 /// Adds a constraint to CS, enforcing that the addition of the allocated numbers in vector `v`
-/// is equal to `sum`.
-///
-/// summation(v) = sum
-#[allow(dead_code)]
-pub(crate) fn popcount<F: PrimeField, CS: ConstraintSystem<F>>(
+/// is equal to the value of the variable, `sum`.
+pub(crate) fn popcount_equal<F: PrimeField, CS: ConstraintSystem<F>>(
     cs: &mut CS,
     v: &[Boolean],
-    sum: &AllocatedNum<F>,
+    sum: Variable,
 ) -> Result<(), SynthesisError> {
-    let mut v_lc = LinearCombination::<F>::zero();
-    for b in v {
-        v_lc = add_to_lc::<F, CS>(b, v_lc, F::ONE)?;
-    }
+    let popcount = popcount_lc::<F, CS>(v)?;
 
-    // (summation(v)) * 1 = sum
+    // popcount * 1 = sum
     cs.enforce(
         || "popcount",
-        |_| v_lc,
+        |_| popcount,
         |lc| lc + CS::one(),
-        |lc| lc + sum.get_variable(),
+        |lc| lc + sum,
     );
 
     Ok(())
 }
 
+/// Adds a constraint to CS, enforcing that the addition of the allocated numbers in vector `v`
+/// is equal to `one`.
+///
+/// summation(v) = one
+#[inline]
+pub(crate) fn enforce_popcount_one<F: PrimeField, CS: ConstraintSystem<F>>(
```

Comment:
No need to change now, but just floating the idea that we should consider changing this to [`enforce_one_hot`](https://en.wikipedia.org/wiki/One-hot). I think that makes the meaning and intent of this even clearer. (I know I was the last person to rename this, just noodling on incremental improvement.)